src/ising.2.cpp

```diff
@@ -0,0 +1,245 @@
+//ising 2.0
+#include <math.h>
+#include <stdlib.h>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <random>
+
+
+using namespace std;
+
+double rnd (){
+    
+    /*http://stackoverflow.com/a/22935167*/
+    random_device rd{};    
+    mt19937 engine{rd()};
+    uniform_real_distribution<double> dist{0.0, 1.0};
+    double x = dist(engine);
+    return x;
+}
+
+void initialize(int N, int *** si){
+    
+    for (int x = 0; x < N; x++){
+        si[x] = new int * [N];
+        
+        for (int y = 0; y < N; y++){
+            si[x][y] = new int [N];
+            
+            for (int z = 0; z < N; z++){
+                int p  = (int) rnd();
+                if (p > 0.5) si[x][y][z] = 1;
+                else si[x][y][z] = -1;
+            }
+        }
+    }
+    
+    return;
+}
+
+int neighbors(int N, int *** si, int * point){
+    
+    int x = point[0];
+    int y = point[1];
+    int z = point[2];
+    
+    int xp = (x+N-1) % N;
+    int xm = (x+1) % N;
+    int yp = (y+N-1) % N;
+    int ym = (y+1) % N;
+    int zp = (z+N-1) % N;
+    int zm = (z+1) % N;
+    
+    int sj = si[xp][y][z] + si[xm][y][z] + si[x][yp][z] + si[x][ym][z] + si[x][y][zp] + si[x][y][zm];
+    
+    return sj;
+}
+
+
+double Magnetization(int N, int *** si){
+    
+    double M = 0.0;
+    
+    for (int i = 0; i < N; i++)
+        for (int j = 0; j < N; j++)
+            for (int k = 0; k < N; k++)
+                M = M+(double)si[i][j][k];
+            
```

Comment:
This part of the program can be optimized. It is costly to calculate the magnetization by scanning `N * N * N` sites after each Monte Carlo step. Instead, we can make use of the fact that only one site flips per step.

The general procedure would be the following:

1.  Calculate and store the total magnetization of the system after initialization by doing a full scan over all lattice sites and store in `M`.
2.  Enter the loop and flip a spin.
3.  Calculate the *change* in magnetization of flipping the single spin. For the Ising model, this is
    ```c++
    deltaM = -2 * si[i][j][k];
    ```
    Thus, the total magnetization of the system *after* flipping the spin would be
    ```c++
    M = M + deltaM;
    ```
4.  Keep the change if the spin-flip passes the test in the Metropolis algorithm.

This will remove the need for the 3 nested for loops that are currently in the code.